### ISSUE TITLE
Auth data is not present on serverside - NextJS App Router

### ISSUE THREAD

**Issue by SimonHylander** [clerk/javascript#1528]
- [x] Review the documentation: https://clerk.com/docs
- [x] Search for existing issues: https://github.com/clerkinc/javascript/issues
- [x] Go through package changelog files.
- [ ]  Provide the Frontend API key from your application dashboard.

## Package + Version

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [x] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`
- [ ] other:

### Dependencies + versions

```json
{
  "name": "app-router",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "build": "next build",
    "dev": "next dev",
    "lint": "next lint",
    "start": "next start"
  },
  "dependencies": {
    "@clerk/nextjs": "^4.23.1",
    "@clerk/themes": "^1.7.5",
    "@t3-oss/env-nextjs": "^0.3.1",
    "dotenv": "^16.3.1",
    "encoding": "^0.1.13",
    "next": "^13.4.12",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "superjson": "1.12.2",
    "svix": "^1.7.0",
    "zod": "^3.21.4"
  }
}
```

### Browser/OS

node 18.16.1

## Description

After logging in, all auth properties are null in the server component using the auth() function.
Auth properties are present using a client component with useAuth().
If I try to log auth in the "afterAuth" middleware all properties are also null.

```jsx
<SignedIn></SignedIn>
``` 
only works in client components.

demo project https://github.com/SimonHylander/next-app-router-clerk

**Comment by dimkl** [comment]
I ran the example application and after sign-in 3 requests were executed to reload the components using RSC. Of the 3 requests, 2 are for the same component. The first 2 requests were executed without session data, but the 3rd was executing (2nd request for the 1 component) is executed with the session populated.

Uploading Screen Recording 2023-08-03 at 13.22.57.mov…
Even though it's not optimal it seems that the demo works. I will take a better look at it and come back to you.

**Comment by clementprevot** [comment]
Hi!

I have noticed the same behavior.
I use `const { userId } = auth()` in a server side component (a layout in this case) and `const { userId } = useAuth()` in a client component (the page at the same level as the layout).

When I authenticate using a magic link, I get redirected to my page with `__clerk_status=verified` and a `__clerk_created_session`, but the log in the server side return a `null` `userId` while in the client component I do have my proper userId.
What is even weirder is that hard-refreshing the page in the browser fixes everything and I have my `userId` in both the server and client component

**Comment by ethansnow2012** [comment]
I also encounter this.
 getAuth() do return userId when I am on the localhost but not when I deployed it to vercel.

**Comment by Indu-Chandana** [comment]
> Hi!
> 
> I have noticed the same behavior. I use `const { userId } = auth()` in a server side component (a layout in this case) and `const { userId } = useAuth()` in a client component (the page at the same level as the layout).
> 
> When I authenticate using a magic link, I get redirected to my page with `__clerk_status=verified` and a `__clerk_created_session`, but the log in the server side return a `null` `userId` while in the client component I do have my proper userId. What is even weirder is that hard-refreshing the page in the browser fixes everything and I have my `userId` in both the server and client component

I have similar issue

**Comment by eob** [comment]
I'm also encountering this bug. Similar to @clementprevot, immediately after login I get a null auth state on the server-side but a non-null auth state on the client side. 

This prevents redirecting to a "post-login" SSR page that performs a few required account-update activities before redirecting the user to their eventual destination

**Comment by jamierumbelow** [comment]
I'm afraid I'm also seeing the same problem. Client side authenticates fine, server returns `null` for everything.

What's weird is that it's working on `localhost` so it's very difficult to debug. Perhaps it's something to do with Vercel's serverless functions?

**Comment by jamierumbelow** [comment]
Okay, I did some more debugging and managed to make it work by deleting the `__client_uat` cookie.

For whatever reason, this cookie wasn't being cleared or replaced correctly when I logged out.

I'm still not sure why:

1. The cookie wasn't reset when I logged out / in
2. This caused this failure

On (2), my suspicion is that it was triggering [an invalid JWT handshake](https://github.com/clerk/javascript/blob/main/packages/backend/src/tokens/handshake.ts#L43), which would explain why I was seeing a `standard-signed-out` reason in my logs. (This is confirmed [here](https://github.com/clerk/javascript/blob/ad7bc7056b7bdc29a19d976e7db0c12053927f26/packages/clerk-js/src/utils/cookies/clientUat.ts#L18)).

**Comment by wowjulian** [comment]
Same issue with Firefox. @BRKalow could you reopen the issue? I believe it's reproducible with Firefox, not Chrome.

**Comment by edgarlcs** [comment]
I had the same issue when I was trying to fetch a POST request from a Server Component to a Route Handler. I could get the auth data in the server component using auth() or currentUser(), but it showed null on the route handler. Turns out doing this is a mistake and you should not be fetching route handlers from a server component as it says in [this video](https://www.youtube.com/watch?v=RBM03RihZVs&t=961s&ab_channel=Vercel) by Vercel.

I guess is not a 'fix' but after changing my logic to the server component I can say the issue is fixed now.

**Comment by edanphillip** [comment]
i was having this same issue on chrome where where my middleware function kept redirecting to the sign-in even though i was authenticated in clerk. also debug mode showed that all auth fields were null. this was the case only in production and didnt happen while i was on localhost. i opened the prod deployment in chrome incognito mode and suprisingly this error was gone and everything worked fine. I stopped being redirected to sign-in even though im logged in. this hinted that the problem was cookies to do with cookies. So, i cleared my cookies for my production domain and then everything worked as originally intended.

tldr; clearing cookies fixed this same issue for me.

**Comment by kellenproctor** [comment]
Clearing the cookies in local development cleared the issue of seeing null returned for auth() and currentUser() in a server component.

**Comment by weisisheng** [comment]
Hmm, tried to clear cookies and same error results.  Both in FFOX and chrome.  @clerk/nextjs@4.29.3

**Comment by luke-edward** [comment]
> i was having this same issue on chrome where where my middleware function kept redirecting to the sign-in even though i was authenticated in clerk. also debug mode showed that all auth fields were null. this was the case only in production and didnt happen while i was on localhost. i opened the prod deployment in chrome incognito mode and suprisingly this error was gone and everything worked fine. I stopped being redirected to sign-in even though im logged in. this hinted that the problem was cookies to do with cookies. So, i cleared my cookies for my production domain and then everything worked as originally intended.
> 
> tldr; clearing cookies fixed this same issue for me.

This was the only thing that helped thanks

**Comment by LukaGiorgadze** [comment]
I had the same error and here's how I got and how I fixed it:

OS: Mac
Browser: Chrome

1. Created production environment by cloning development environment;
2. Tried to log in using Google Social login and did not work, it requires some configuration (but cookies are kept in my browser);
3. Configured Google auth, and when I tried to log in, I was experiencing the same issue - `auth` did not work on server side.

I just removed cookies and all application data, including cache and it worked. Try this.

**Comment by BinaryWizardX** [comment]
Hi !

I am also having the same issue on my local development environment. I works fine on the client side components and return null in serverside components. I've tried clearing cookies and cache data but the problem still exists. If somebody found any solutions to this issue please reply? 

Thank you!

**Comment by DoctorLimitless** [comment]
why would you close this issue if its not solved?!!

**Comment by weisisheng** [comment]
Just to add to the discussion, I turned on my sync with server for time/data, cleared cookies, tried with VPN on and off, set firefox to accept cookies from session locally (0.0.0.0:3000 localhost:3000), still get the cookie “__client_uat” = 0 error.

Ubuntu 24.04,  @clerk/clerk-js@5.5.3, @clerk/nextjs@5.1.3, node v20.11.0, nextjs@14.23.0

**Comment by rlking1985** [comment]
Same issue, can't access auth() within an api route in next js.

**Comment by GGalupo** [comment]
Same issue here.

**Comment by ozair-dev** [comment]
Same issue here. Works fine  in localhost. But when hosted on vercel or any other hosting platform, the auth is not presend on serverside (and user is redirected to sign in page). But in client side, I can see the signedin user with `useAuth`

@GGalupo , @rlking1985 , are you facing this issue in localhost? or only in hosted version?

**Comment by elnfar** [comment]
I spent the whole day tryinna fixing it, it looks like if you are using route handler in app route. app/api/route.ts currentUser or auth() always returns back null. Instead, you can use server actions directly in Nextjs which seems to be working fine. Otherwise, you can still define and pass next in your middleware/ header and cookie.

**Comment by QingjiaTsang** [comment]
> Hi ! 你好！
> 
> I am also having the same issue on my local development environment. It works fine on the client side components and return null in serverside components. I've tried clearing cookies and cache data but the problem still exists. If somebody found any solutions to this issue please reply?我在本地开发环境中也遇到了同样的问题。它在客户端组件上工作正常，在服务器端组件中返回 null。我尝试清除cookie和缓存数据，但问题仍然存在。如果有人找到此问题的任何解决方案，请回复？
> 
> Thank you! 谢谢！

Totally same problem here. How do you figure it out finally?

**Comment by QingjiaTsang** [comment]
> Hi! 你好！
> 
> I have noticed the same behavior.我注意到了同样的行为。 I use `const { userId } = auth()` in a server side component (a layout in this case) and `const { userId } = useAuth()` in a client component (the page at the same level as the layout).我在服务器端组件（在本例中为布局）和 `const { userId } = useAuth()` 客户端组件（与布局处于同一级别的页面）中使用 `const { userId } = auth()` 。
> 
> When I authenticate using a magic link, I get redirected to my page with `__clerk_status=verified` and a `__clerk_created_session`, but the log in the server side return a `null` `userId` while in the client component I do have my proper userId.当我使用魔术链接进行身份验证时，我被重定向到我的页面 `__clerk_status=verified` `__clerk_created_session` ，但服务器端的登录在客户端组件中返回了一段时间 `null` `userId` ，我确实有正确的 userId。 What is even weirder is that hard-refreshing the page in the browser fixes everything and I have my `userId` in both the server and client component更奇怪的是，在浏览器中硬刷新页面可以修复所有问题，并且我在服务器和客户端组件中都有我的 `userId`

so how's it going now? It seems like it's still not working for nextjs api handler.

**Comment by QingjiaTsang** [comment]
It works when getting userId out of auth() from CSC & SSC, as well as making requests from the client side to the API route handler where I use the auth function to get userId. I get the nullish value of userId, only when I run the fetch function from the SSC to the API route handler. I guess that's maybe because SSC loses the context of the clerk auth when making data fetching. 

So make sure making the request from client component.

**Comment by nandoviski** [comment]
I'm having the same issue, I'm using NextJs 13 but using Pages, not the app directory, I'm also using tRpc, on the client side I'm logged in and can see the user, but in the server side, useAuth returns null for everything, I'm doing what they said in the docs, but nothing seems to help

I'd like to know if there's plans for this to be fixed or any work around? we are planning to migrate the authentication in our website, but with this bug would be impossible to recommend Clerk, can't even test it

**Comment by elnfar** [comment]
I found out that it's not only Clerk, Next-auth also does the same, please sure you are not sending the requet from SSC to api route directly, it should be from client component to api route, then it works. For just trying you can create a GET function in api/test/route.ts and get the user using auth(), it should work with GET in api route.

**Comment by ozair-dev** [comment]
> but in the server side, useAuth returns null for everything

@nandoviski I don't think that `useAuth` is supposed to work in server side. You should use `auth()` on server side to get the authenticated user. Or maybe `getAuth()`

**Comment by nandoviski** [comment]
@ozair-dev sorry, I mean getAuth() from "@clerk/nextjs/server", I'll edit my post

**Comment by nandoviski** [comment]
@elnfar thanks, I tried but didn't work, I login in the website, and called the endpoint, but it returns null, maybe I'm doing something wrong, but I doing everything the say in the docs
<img width="1215" alt="Screenshot 2024-06-25 at 9 19 27 PM" src="https://github.com/clerk/javascript/assets/12448371/f1286a9c-6a2d-4204-9d2f-8e9694c67670">

**Comment by elnfar** [comment]
> @elnfar thanks, I tried but didn't work, I login in the website, and called the endpoint, but it returns null, maybe I'm doing something wrong, but I doing everything the say in the docs <img alt="Screenshot 2024-06-25 at 9 19 27 PM" width="1215" src="https://private-user-images.githubusercontent.com/12448371/342708513-f1286a9c-6a2d-4204-9d2f-8e9694c67670.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMTYwNTEsIm5iZiI6MTcxOTMxNTc1MSwicGF0aCI6Ii8xMjQ0ODM3MS8zNDI3MDg1MTMtZjEyODZhOWMtNmEyZC00MjA0LTlkMmYtOGU5Njk0YzY3NjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDExNDIzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNmEzZGYxOTU4ZjQ1NjgzMDgwNjM0ZGQxN2JkZGQzZjIwZWZiMTZkOTVhNjJhYmQ0MGU1ZTE1NTZlMTI0MzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VY-XbIq3iA-E61qsTfCwXWillTP5Z-Tf1xJpzMSmO5M">

Just, are you sure you are passing data back to server after the user is singed in or up? You can also log your request to see what to do you have there, and make sure you have clean config in middleware.ts for Clerk ( it changed after the update ).

**Comment by nandoviski** [comment]
@elnfar the middleware is what I got from the docs, `export default clerkMiddleware();` and the config, I can log the request but I'm not sure what I need to have there... I'll see if I can find out... thanks again

**Comment by umran** [comment]
There doesn't seem to be a definitive explanation as to why this is happening. Why does this issue arise only in production while the exact same code works in dev?

**Comment by mateussmohamed** [comment]
Same issue here... `clerkMiddleware();` is losing context when navigating across pages in my application.

**Comment by jescalan** [comment]
Hi friends! Is anyone here able to provide a reproduction we could look into? We hear that this issue is impacting a lot of people, but it's really hard for us to investigate without being able to reproduce 😅

**Comment by LukaGiorgadze** [comment]
> Hi friends! Is anyone here able to provide a reproduction we could look into? We hear that this issue is impacting a lot of people, but it's really hard for us to investigate without being able to reproduce 😅
> 
> In case it's helpful context at all, this will happen if your server is set up on a different subdomain from your client-side app, and we have seen plenty of instances were users run into this problem for that reason - details on how to resolve [can be found here](https://clerk.com/docs/backend-requests/making/cross-origin#using-fetch-with-clerk-js).
> 
> If anyone is not able to make a reproduction but would be willing to get on a call with me and screenshare, I am sure we could pin it down pretty quickly as well. You can shoot me an email at `jeff` at company name dot com.

https://github.com/clerk/javascript/issues/1528#issuecomment-2077683692

**Comment by umran** [comment]
@jescalan in my case, it turned out i wasn't using the right credentials (client attempted to connect with sandbox keys in production). I suspect a lot of people are running into this issue via that route. but it sounds like there maybe a bug that causes this behaviour too even if the environment is set up correctly.

**Comment by jescalan** [comment]
@umran Helpful context here, thank you! Using dev keys in production is definitely a problem that we see a lot of customers run into. Though since we have shipped the [development mode ui changes](https://clerk.com/changelog/2024-08-02-dev-notice), it's been substantially less 😌

**Comment by Wolfleader101** [comment]
For anyone still having this issue, `auth()` now returns a Promise<Auth`

So your code will now need to look like

```ts
 const { userId } = await auth();
```

**Comment by Leonardo-Mazzuca** [comment]
Hi! I'm passing fot this same issue so far. I'm using Next 15


import { SignInButton, SignedIn, SignedOut, UserButton, useUser } from "@clerk/nextjs";
import { auth } from "@clerk/nextjs/server";

const Header = async () => {

    //here is null
    const {userId} = await auth();
    
    // console.log(user);
    console.log(userId);
    

  return (
    <nav className="navbar">
      <div className="navbar-container">
        <h2>Expense Tracker</h2>
    
        <div className="">
          <SignedOut>
            <SignInButton />
          </SignedOut>

          <SignedIn>
            <UserButton />
          </SignedIn>
        </div>
      </div>
    </nav>
  );
};

export default Header;

I'm watching a travery media video, this one: https://youtu.be/I6DCo5RwHBE?si=TqeJAFVmG2hjKtR8

And i'm following exactly what he is doing in the video, but, it doesn't working for me. 
- The auth( ) helper is returning null
- The currentUser ( ) it's null as well
- i've tried do route handlers, but, it fails as well


So far i've been facing a lot of problems with this using the Clerk, someone have figure out any solution????
I can show my code repo for tests if needed

**Comment by rajenderK7** [comment]
So what I understood is we get the nullish value from auth() and currentUser() functions when we're making a request from a server side component to Next's api routes. And the easiest solution is to make the component a client side one. But what if there's a genuine use case to have a server component and not a client side one. In that case we don't have a solution at all?

**Comment by BrannJoly** [comment]
> So what I understood is we get the nullish value from auth() and currentUser() functions when we're making a request from a server side component to Next's api routes. And the easiest solution is to make the component a client side one. But what if there's a genuine use case to have a server component and not a client side one. In that case we don't have a solution at all?

Unfortunately there are things that can't run on the client side (namely the middleware)
I'm a little bit shocked this as been closed as "not planned" eventhough the root cause has not been found and some users are still experiencing the issue ...

**Comment by jescalan** [comment]
Just a note for anyone looking on - Brann reached out to Clerk support (which is the recommendation, as our support team does not monitor github issues, so thank you), and we discovered that an `Authorization` header was being passed with requests that didn't contain a Clerk JWT, which caused the server to return an error as if that header is utilized, it's expected to contain a Clerk auth value ([relevant docs here](https://clerk.com/docs/backend-requests/making/cross-origin)).

If anyone else is having issues with their Clerk app, please [reach out to our support team](https://clerk.com/contact/support). Auth data does work on the server side, this is not a known or open issue, and it's not "not planned" to fix it. I corrected the issue status to account for this and have locked further posts to avoid confusion.
